File: KERN_SRC/ftl/dftl.c
#include "ftl.h"

#include <linux/delay.h>
#include <linux/list.h>

typedef struct {
  u32 lpn[PAGE_PER_BLOCK];
  u32 id;
  u32 clean;
  u32 dirty;
  struct list_head list;
} dftl_block;
dftl_block dftl_blocks[BLOCK_NUMBER];

LIST_HEAD(free_block);
LIST_HEAD(gc_block);

struct {
  u32 ppn;
  u32 block;
  bool cached;
  bool dirty;
} dftl_table[TABLE_SIZE];

struct {
  u32 lpn;
} dftl_cache[CACHE_SIZE];

u32 head = 0;
u32 tail = 0;

u32 next(u32 point) {
  u32 next = point + 1;
  if ( next >= CACHE_SIZE ) {
    next = 0;
  }
  return next;
}

void cache(u32 lpn) {
  u32 next_head = next(head);

  dftl_table[lpn].cached = true;

  if ( next_head == tail ) {
    u32 evict_lpn = dftl_cache[tail].lpn;
    dftl_table[evict_lpn].cached = false;
    if ( dftl_table[evict_lpn].dirty ) {
      udelay(FLASH_PAGE_READ_DELAY);
      udelay(FLASH_PAGE_WRITE_DELAY);
      dftl_table[evict_lpn].dirty = false;
    }
    tail = next(tail);
  }

  dftl_cache[head].lpn = lpn;
  head = next_head;
}

void dftl_init(void) {
  int i = 0;
  for(i = 0; i < TABLE_SIZE; i++) {
    dftl_table[i].ppn = PAGE_NUMBER;
    dftl_table[i].cached = false;
    dftl_table[i].dirty = false;
  }

  for(i = 0; i < BLOCK_NUMBER; i++) {
    list_add(&(dftl_blocks[i].list), &(free_block));
    dftl_blocks[i].id = i;
    dftl_blocks[i].clean = PAGE_PER_BLOCK;
    dftl_blocks[i].dirty = 0;
  }
  return;
}

u32 dftl_read(u32 lpn) {
  if ( !dftl_table[lpn].cached ) {
    udelay(FLASH_PAGE_READ_DELAY);
    cache(lpn);
  }

  return dftl_table[lpn].ppn;
}

void dftl_gc(void) {

}

dftl_block * dftl_free_block(void) {
  dftl_block * block = NULL;
  while(block == NULL){
    if( list_empty(&free_block) ) {
      dftl_gc();
    }

    if( !list_empty(&free_block) ){
      block = list_first_entry(&free_block, dftl_block, list);
    }
    if( block != NULL && block->clean == 0 ) {
      list_del(&block->list);
      block = NULL;
    }
  }
  return block;
}

u32 dftl_write(u32 lpn) {
  u32 ppn = 0;
  dftl_block * block;

  if( !dftl_table[lpn].cached ) {
    udelay(FLASH_PAGE_READ_DELAY);
    cache(lpn);
  }

  if( !( dftl_table[lpn].ppn >= PAGE_NUMBER ) )  {
    int block_id = dftl_table[lpn].ppn / PAGE_PER_BLOCK;
    dftl_blocks[block_id].dirty += 1;
  }

  block = dftl_free_block();

  ppn = block->id * PAGE_PER_BLOCK + PAGE_PER_BLOCK - block->clean;
  block->clean -= 1;
  dftl_table[lpn].ppn = ppn;
  dftl_table[lpn].block = block->id;
  dftl_table[lpn].cached = true;
  dftl_table[lpn].dirty = true;

  return ppn;
}

struct ftl dftl = {
  dftl_init,
  dftl_read,
  dftl_write
};

File: USER_SRC/scala/src/main/scala/org/janzhou/ftl/DFTL.scala
package org.janzhou.ftl

class DFTL(device:Device) extends FTL(device) {

  println("DFTL")

  class dftl_block(val id:Int) {
    var lpns = List[Int]()
    var clean:Int = device.PagesPerBlock
    var dirty:Int = 0
  }

  protected val dftl_blocks = for( i <- 0 to device.NumberOfBlocks - 1 ) yield {
    new dftl_block(i)
  }
  protected var free_block = dftl_blocks.toList
  protected var gc_block = List[dftl_block]()
  protected var full_block = List[dftl_block]()

  class dftl_mapping_entry(val lpn:Int) {
    var block:dftl_block = null
    var ppn:Int = device.TotalPages
    var cached:Boolean = false
    var dirty:Boolean = false
  }

  protected val dftl_table = for ( i <- 0 to device.TotalPages - 1 ) yield {
    val block_id = i / device.PagesPerBlock
    new dftl_mapping_entry(i)
  }
  protected var dftl_cache = List[dftl_mapping_entry]()

  protected def clean_cache = {
    while( dftl_cache.length > device.CacheSize ) {
      val cache = dftl_cache.head

      if ( cache.dirty ) {
        device.PageReadDelay
        device.PageWriteDelay

        dftl_cache = dftl_cache.drop(1)
      }
    }
  }

  def read(lpn:Int):Int = {
    if ( dftl_table(lpn).cached == false ) {
      dftl_table(lpn).cached = true
      dftl_table(lpn).dirty = false

      dftl_cache = dftl_cache :+ dftl_table(lpn)

      device.PageReadDelay
    }

    device.PageReadDelay //data
    dftl_table(lpn).ppn
  }

  protected def move = {
    val block = {
      val candidate_blocks = full_block

      var gc = candidate_blocks.head
      for ( b <- candidate_blocks ) {
        if ( b.dirty > gc.dirty ) {
          gc = b
        }
      }
      gc
    }

    for ( lpn <- block.lpns ) {
      val from = read(lpn)
      val to = write_withoutGC(lpn)
      device.move(from, to)
    }
  }

  def gc = {
    if ( gc_block.isEmpty ) {
      move
    }

    val block = gc_block.head
    block.lpns = List[Int]()
    block.clean = device.PagesPerBlock
    block.dirty = 0
    gc_block = gc_block.drop(1)
    free_block = free_block :+ block

    device.BlockEraseDelay
  }

  def trim(lpn:Int):Unit = {
    if ( dftl_table(lpn).block != null ) {
      if ( dftl_table(lpn).cached == false ) {
        device.PageReadDelay
      }

      val block = dftl_table(lpn).block
      block.lpns = block.lpns.filter(_ != lpn)
      block.dirty += 1

      if ( block.dirty >= device.PagesPerBlock ) {
        gc_block = gc_block :+ block
        full_block = full_block.filter( _ != block )
      }

      dftl_table(lpn).block = null
      dftl_table(lpn).ppn = device.TotalPages
      dftl_table(lpn).cached = true
      dftl_table(lpn).dirty = true

      dftl_cache = dftl_cache :+ dftl_table(lpn)
    }
  }

  def write(lpn:Int):Int = {
    if( free_block.length <= 2 ) gc
    write_withoutGC(lpn)
  }

  private def write_withoutGC(lpn:Int):Int = {
    val block = free_block.head

    if ( dftl_table(lpn).block != null ) trim(lpn)

    val ppn = ( block.id + 1 ) * device.PagesPerBlock - block.clean
    block.clean -= 1
    if( block.clean <= 0 ) {
      free_block = free_block.drop(1)
      full_block = full_block :+ block
    }
    block.lpns = block.lpns :+ lpn

    dftl_table(lpn).block = block
    dftl_table(lpn).ppn = ppn

    device.PageWriteDelay //data
    ppn
  }
}
